lib/SILOptimizer/Transforms/SILMem2Reg.cpp

```diff
@@ -980,15 +980,21 @@ void StackAllocationPromoter::fixBranchesAndUses(BlockSetVector &phiBlocks,
 /// insert the end into them.  Find those blocks by visiting the successors of
 /// each new lexical begin_borrow's block.  End the borrow scope in the visited
 /// blocks if any of the following conditions are met:
-/// (1) The block is terminal--if we haven't managed to end the borrow scope
-///     before the last block in the function, it must be ended there.
-/// (2) Any of its successors don't have live-in values for the alloc_stack--in
-///     order for a block to propagate the copied value so that it might
-///     eventually be destroyed, the block must pass that value on to all its
-///     successors; if any successor does not take that value, then neither it
-///     nor any of its successors could possibly destroy the value.
-/// (3) The block had a dealloc_stack--if the allocated space was deallocated
-///     in this block, the lifetime of any value stored in it must end there.
+/// (1) The block is terminal.
+///
+///     If we haven't managed to end the borrow scope before the last block in
+///     the function, it must be ended there.
+/// (2) Any of its successors don't have live-in values for the alloc_stack.
+///
+///     In order for a block to propagate the borrowed value so that it might
+///     eventually be end_borrow'd, the block must pass that value on to all its
+///     successors; if any successor does not take that value, then neither that
+///     successor nor any of its own successors could possibly destroy the
+////    value.
+/// (3) The block had a dealloc_stack.
+///
+///     If the allocated space was deallocated in this block, the lifetime of
+///     any value stored in it must end there.
```

Comment:
I think we should clarify the requirements of this algorithm vs the implementation. The lexical scope needs to end at any of
1. dealloc_stack
2. function exit
3. iterated dominance frontier relative to the points where the alloc_stack is initialized

Note that loaded value lifetimes and destroy_value uses should have no effect.

Only case #3 is interesting. We can simplify the problem by noting that it is valid to end the lexical lifetime anywhere between the stack location's deinitialization and its dealloc_stack. 

We should note somewhere that it is illegal for any SIL pass to convert this
```
%v = load [copy] %a
...<stuff>
destroy %v
destroy_addr %a
```

into
```
%v = load [take] %a
...<stuff>
destroy %v
```

Unless the pass proves that the destroy can safely be hoisted above <stuff> according the same logic that lets us shrink borrow scopes.

Then we can reason that all paths that have a lexical borrow must either end in a function exit or destroy_addr (or [take]).

We may still need to be careful about dominance though. Can we handle this case?

```
bb1
  br bb5

bb2:
  store [init]
  cond_br bb3, bb4

bb3:
  destroy_addr

bb4:
  br bb5

bb5:
  unreachable
```